### PR TITLE
*: add checklist

### PR DESCRIPTION
**Pull request by coocood**
The checklist lists some items to check before sending a PR.
Following the checklist saves reviewers' time, gets PR reviewed and merged faster.

Print the checklist in `make dev` or `make test` to remind the PR author.

The more items may be added when we find something is important but gets ignored frequently.

### CODE REVIEW

**Comment by hanfei1991** [comment]
LGTM

File: checklist.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+# Self Review
+Have you reviewed every line of your changes by yourself?
+# Test
+Have you added enough test case to cover the new feature or bug fix?
+Also, add comments to describe your test cases.
+# Naming
+Do function names keep consistent with its behavior?
+Is it easy to infer the function's behavior by its name?
+# Comment
```
Comment (zhexuany) [inline checklist.md]: make goword.

**Comment by zhexuany** [comment]
@coocood A better practice is of putting a PR template in `.github` instead of putting it in Makefile.

**Comment by coocood** [comment]
@ngaut @siddontang @shenli PTAL

File: checklist.md (hunk 1 above)
Comment (coocood) [inline checklist.md]: It's included in hooks already.

**Comment by coocood** [comment]
@zhexuany 
I think it's too late to remind those important things when you already pushed your branch and about to send a PR.

Put it in Makefile makes it exposed to people more frequently, and when you are waiting for compile, you can utilize this time to do checks.

When we get too many checklist items, we can only display a subset of them randomly, which can not be achieved by the template.

File: checklist.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+Is there any code can confuse the reviewer?
```
Comment (tiancaiamao) [inline checklist.md]: s/can/that would

File: checklist.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Add comments on them! You'll be asked to do so anyway.
+Make sure there is no syntax or spelling error in your comments.
+Some online syntax checking tool like Grammarly may be helpful.
+# Refactor
+Is there any way to refactor the code to make it more readable?
+If the refactoring touches a lot of existing code, send another PR to do it.
+# Single Purpose
+Make sure the PR does only one thing and nothing else.
+# Diff Size
+Make sure the diff size no more than 500, split it to small PRs if it is too large.
```
Comment (tiancaiamao) [inline checklist.md]: s/split it to/split it into

File: checklist.md (hunk 2 above)
Comment (Wenting0905) [inline checklist.md]: Is there any code that confuses reviewers?

File: checklist.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+Have you added enough test case to cover the new feature or bug fix?
+Also, add comments to describe your test cases.
+# Naming
+Do function names keep consistent with its behavior?
+Is it easy to infer the function's behavior by its name?
+# Comment
+Is there any code can confuse the reviewer?
+Add comments on them! You'll be asked to do so anyway.
+Make sure there is no syntax or spelling error in your comments.
+Some online syntax checking tool like Grammarly may be helpful.
```
Comment (Wenting0905) [inline checklist.md]: tool --> tools

File: checklist.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+# Refactor
+Is there any way to refactor the code to make it more readable?
+If the refactoring touches a lot of existing code, send another PR to do it.
```
Comment (Wenting0905) [inline checklist.md]: 1. touches --> involves/concerns/influences. What does "touch" mean here? Related to the codes or have an impact on the existing codes? 
2. code --> codes
3. send a PR or create a PR?

File: checklist.md (hunk 3 above)
Comment (Wenting0905) [inline checklist.md]: Make sure the diff size is no more than 500; if it is too large, split it into smaller PRs.
BTW, does the diff size have a unit? The number 500 seems strange without a unit.

File: checklist.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Following the checklist saves the reviewers' time and gets your PR reviewed faster.
+
+# Self Review
+Have you reviewed every line of your changes by yourself?
+# Test
+Have you added enough test case to cover the new feature or bug fix?
```
Comment (Wenting0905) [inline checklist.md]: same with XuHuaiyu

**Comment by Wenting0905** [comment]
Rest LGTM

**Comment by Wenting0905** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**